File: catroid/src/test/java/org/catrobat/catroid/content/actions/FaceNameDetectActionTest.java
package org.catrobat.catroid.content.actions;

import org.catrobat.catroid.FaceRecognizer.FaceDetector;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;

import java.lang.reflect.Field;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;

/**
 * The detect brick's action.
 *
 * There is no camera in a unit test, so every detection is refused at the
 * permission check and the callback fires immediately. That makes the state
 * machine fully deterministic here, which is the part worth testing.
 *
 * The case that matters most is secondCallStartsAFreshDetection. The Restart
 * button re-runs the script, and if the action remembers that it already
 * finished, it returns true straight away and the camera never opens. That is
 * the bug this suite exists to catch.
 */
@RunWith(RobolectricTestRunner.class)
public class FaceNameDetectActionTest {

	private FaceNameDetectAction action;

	@Before
	public void setUp() {
		FaceDetector.resetForNewRun();
		action = new FaceNameDetectAction();
	}

	@After
	public void tearDown() {
		FaceDetector.resetForNewRun();
	}

	// ---------------- One pass ----------------

	@Test
	public void theActionFinishes() {
		// With no camera the detection resolves at once, so act returns true.
		assertTrue(action.act(0.016f));
	}

	@Test
	public void aFinishedActionKeepsReturningTrueWithinTheSamePass() {
		assertTrue(action.act(0.016f));
		// The sequence has moved on, but a stray call must not restart anything
		// halfway through. It reports done, then the next call is a new pass.
		assertTrue(action.act(0.016f));
	}

	@Test
	public void theActionAlwaysHasAName() {
		action.act(0.016f);
		assertNotNull(action.getDetectedName());
		assertFalse(action.getDetectedName().isEmpty());
	}

	@Test
	public void withNoCameraTheNameIsUnknownRatherThanNull() {
		action.act(0.016f);
		assertEquals(FaceDetector.UNKNOWN, action.getDetectedName());
	}

	// ---------------- Restart, the bug ----------------

	/**
	 * Press Restart, the script runs again, the brick is reached again. The
	 * camera must open again.
	 *
	 * Proof that begin() ran a second time: it calls resetForNewRun, which wipes
	 * the stored name. If the action had simply returned true without doing any
	 * work, the planted name would still be there.
	 */
	@Test
	public void secondCallStartsAFreshDetection() {
		assertTrue(action.act(0.016f));

		// Pretend a real detection had stored a name.
		setDetectorName("salah");
		setDetectorDone(true);

		assertTrue("the brick must complete again", action.act(0.016f));

		assertEquals("the second pass did not run a fresh detection",
				FaceDetector.UNKNOWN, FaceDetector.getLastName());
	}

	@Test
	public void manyPassesEachRunAFreshDetection() {
		for (int pass = 1; pass <= 5; pass++) {
			setDetectorName("stale" + pass);
			setDetectorDone(true);

			assertTrue("pass " + pass + " did not complete", action.act(0.016f));
			assertEquals("pass " + pass + " reused the old name",
					FaceDetector.UNKNOWN, FaceDetector.getLastName());
		}
	}

	@Test
	public void restartClearsTheAction() {
		action.act(0.016f);
		action.restart();
		assertEquals(FaceDetector.UNKNOWN, action.getDetectedName());
		assertTrue(action.act(0.016f));
	}

	@Test
	public void resetClearsTheAction() {
		action.act(0.016f);
		action.reset();
		assertEquals(FaceDetector.UNKNOWN, action.getDetectedName());
		assertTrue(action.act(0.016f));
	}

	@Test
	public void aReusedActionObjectBehavesLikeAFreshOne() {
		// Catroid may reuse the action object rather than building a new one.
		action.act(0.016f);
		action.act(0.016f);
		action.restart();

		setDetectorName("salah");
		setDetectorDone(true);

		assertTrue(action.act(0.016f));
		assertEquals(FaceDetector.UNKNOWN, FaceDetector.getLastName());
	}

	// ---------------- Independence ----------------

	@Test
	public void twoActionsDoNotShareState() {
		FaceNameDetectAction other = new FaceNameDetectAction();

		assertTrue(action.act(0.016f));
		assertTrue(other.act(0.016f));

		assertNotNull(action.getDetectedName());
		assertNotNull(other.getDetectedName());
	}

	@Test
	public void theActionNeverThrows() {
		for (int i = 0; i < 50; i++) {
			action.act(0.016f);
		}
		action.restart();
		action.reset();
		action.act(0f);
		action.act(-1f);
		action.act(1000f);
	}

	// ---------------- Helpers ----------------

	private static void setDetectorName(String name) {
		set("lastName", name);
	}

	private static void setDetectorDone(boolean done) {
		try {
			Field f = FaceDetector.class.getDeclaredField("detectionDone");
			f.setAccessible(true);
			f.setBoolean(null, done);
		} catch (Exception e) {
			throw new AssertionError(e);
		}
	}

	private static void set(String name, Object value) {
		try {
			Field f = FaceDetector.class.getDeclaredField(name);
			f.setAccessible(true);
			f.set(null, value);
		} catch (Exception e) {
			throw new AssertionError(e);
		}
	}
}

File: catroid/src/test/java/org/catrobat/catroid/content/actions/FaceNameDetectActionTest.kt
/*
 * Catroid: An on-device visual programming system for Android devices
 * Copyright (C) 2010-2026 The Catrobat Team
 * (<http://developer.catrobat.org/credits>)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * An additional term exception under section 7 of the GNU Affero
 * General Public License, version 3, is available at
 * http://developer.catrobat.org/license_additional_term
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package org.catrobat.catroid.content.actions

import android.Manifest
import android.app.Application
import androidx.test.core.app.ApplicationProvider
import org.catrobat.catroid.FaceRecognizer.FaceDetector
import org.catrobat.catroid.formulaeditor.SensorHandler
import org.catrobat.catroid.formulaeditor.Sensors
import org.junit.After
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.Shadows.shadowOf

/**
 * The recognition half of the feature: the Face Name Detect brick's action,
 * through FaceDetector, into the face-name sensor that formulas read.
 *
 * Only the camera capture is replaced (FaceDetector.captureStarter). The
 * permission check, the per-run gate, FaceDetector.finish() writing the sensor
 * and the action's waiting logic are the production code.
 */
@RunWith(RobolectricTestRunner::class)
class FaceNameDetectActionTest {

    private lateinit var application: Application
    private lateinit var capture: FakeCapture
    private lateinit var action: FaceNameDetectAction

    @Before
    fun setUp() {
        application = ApplicationProvider.getApplicationContext()
        capture = FakeCapture()
        FaceDetector.captureStarter = capture
        FaceDetector.resetForNewRun()
        action = FaceNameDetectAction()
    }

    @After
    fun tearDown() {
        capture.completePendingWith(null)
        FaceDetector.useCameraCapture()
        FaceDetector.resetForNewRun()
    }

    @Test
    fun recognisedNameIsWrittenIntoTheSensor() {
        grantCamera()

        assertFalse("The action must wait while the camera is working", action.act(FRAME))
        assertEquals(1, capture.starts)

        capture.completePendingWith("Person A")

        assertTrue(action.act(FRAME))
        assertEquals("Person A", action.detectedName)
        assertEquals("Person A", faceNameSensor())
    }

    @Test
    fun actionKeepsWaitingUntilTheCaptureReports() {
        grantCamera()

        assertFalse(action.act(FRAME))
        assertFalse(action.act(FRAME))
        assertFalse(action.act(FRAME))
        assertEquals("Only one capture may be started per brick run", 1, capture.starts)
        assertEquals(FaceDetector.UNKNOWN, faceNameSensor())

        capture.completePendingWith("Person B")

        assertTrue(action.act(FRAME))
        assertEquals("Person B", faceNameSensor())
    }

    @Test
    fun unrecognisedFaceWritesUnknown() {
        grantCamera()

        action.act(FRAME)
        capture.completePendingWith(null)

        assertTrue(action.act(FRAME))
        assertEquals(FaceDetector.UNKNOWN, action.detectedName)
        assertEquals(FaceDetector.UNKNOWN, faceNameSensor())
    }

    @Test
    fun nameFromThePreviousRunIsClearedBeforeTheNextDetection() {
        grantCamera()
        action.act(FRAME)
        capture.completePendingWith("Person A")
        assertTrue(action.act(FRAME))
        assertEquals("Person A", faceNameSensor())

        // The brick is reached again, e.g. in a loop.
        assertFalse(action.act(FRAME))
        assertEquals(
            "A stale name must not be visible while the next detection runs",
            FaceDetector.UNKNOWN,
            faceNameSensor()
        )
        assertEquals(2, capture.starts)

        capture.completePendingWith("Person C")
        assertTrue(action.act(FRAME))
        assertEquals("Person C", action.detectedName)
        assertEquals("Person C", faceNameSensor())
    }

    @Test
    fun missingCameraPermissionFinishesWithUnknownWithoutOpeningTheCamera() {
        SensorHandler.setFaceNameRecognitionResult("Person A")

        assertTrue(action.act(FRAME))

        assertEquals(0, capture.starts)
        assertEquals(FaceDetector.UNKNOWN, action.detectedName)
        assertEquals(FaceDetector.UNKNOWN, faceNameSensor())
    }

    @Test
    fun captureThatNeverReportsTimesOutWithUnknown() {
        grantCamera()

        assertFalse(action.act(FRAME))
        assertTrue("The action must give up after its timeout", action.act(TIMEOUT_PLUS_ONE))

        assertEquals(FaceDetector.UNKNOWN, action.detectedName)
        assertEquals(FaceDetector.UNKNOWN, faceNameSensor())
    }

    private fun grantCamera() {
        shadowOf(application).grantPermissions(Manifest.permission.CAMERA)
    }

    private fun faceNameSensor(): Any? {
        val value = SensorHandler.getSensorValue(Sensors.ON_DEVICE_FACE_RECOGNITION)
        assertNotNull(value)
        return value
    }

    /** Stands in for the camera session; the test decides when and what it reports. */
    private class FakeCapture : FaceDetector.CaptureStarter {
        var starts = 0
            private set
        private var pending: FaceDetector.Callback? = null

        override fun start(context: android.content.Context, onResult: FaceDetector.Callback) {
            starts++
            pending = onResult
        }

        fun completePendingWith(name: String?) {
            val callback = pending ?: return
            pending = null
            callback.onFinished(name, if (name == null) 0f else 0.8f)
        }
    }

    private companion object {
        const val FRAME = 1f / 60f

        /** FaceNameDetectAction.TIMEOUT_SECONDS is 30. */
        const val TIMEOUT_PLUS_ONE = 31f
    }
}
